Api save whitelist

will-moore opened the pull request:
# What this PR does

Restricts Save (PUT or POST) to a "whitelist" of supported object types.

In the ```SaveView``` class that handles PUT and POST, we use a whitelist of supported types ```Project, Dataset, Screen```.
In the ```ObjectView``` subclasses that handle DELETE of various objects we use a CAN_DELETE attribute that is True is the base class and False in subclasses where Delete is NOT supported ```Plate, Image, Well```.

# Testing this PR

 - Existing tests confirm that we can create, update & delete Project, Dataset & Screen. (Plate removed from this test).
 - Added tests that checks Image, Plate & other types cannot be created, updated or deleted.

# Related reading

 Discussed at web meeting on [2017-01-24](https://docs.google.com/document/d/18ep65grhBURv6CcZLGVmvKQvP3RjfuyIhE288AAuFTM/edit#heading=h.axbzbuq0r8pq)

will-moore committed:
Basic whitelist for allowing 'save' objects

will-moore committed:
SaveView restricts PUT to can_put whitelist

chris-allan commented on components/tools/OmeroWeb/omeroweb/api/views.py:
```diff
@@ -392,19 +392,37 @@ class SaveView(View):
     POST to create a new Object and PUT to replace existing one.
     """
 
+    can_put = ['Project', 'Dataset', 'Screen']
```
As these are constants, let's uppercase them.

chris-allan reviewed (changes requested):
We should add some test cases to ensure that the `PUT` and `POST` restrictions are accurate.

will-moore committed:
Uppercase CAN_PUT, CAN_POST. Both allow P/D/S

will-moore committed:
Add CAN_DELETE check into ObjectView.delete()

will-moore committed:
Return status=405 'method not supported' for disallowed methods

will-moore committed:
Add test to check POST, PUT & DELETE not supported for Images

will-moore committed:
flake8 fixes

snoopycrimecop commented:
Conflicting PR. Removed from build [OMERO-DEV-merge-push#577](https://ci.openmicroscopy.org/job/OMERO-DEV-merge-push/577/). See the [console output](https://ci.openmicroscopy.org/job/OMERO-DEV-merge-push/577/consoleText) for more details.
Possible conflicts:
  - PR #5047 will-moore 'Api wells'
    - components/tools/OmeroWeb/test/integration/test_api_images.py
  - PR #5058 will-moore 'Api child count'
    - components/tools/OmeroWeb/test/integration/test_api_images.py

will-moore committed:
Remove 'Plate' from test_container_crud()

snoopycrimecop commented:
Conflicting PR. Removed from build [OMERO-DEV-merge-push#579](https://ci.openmicroscopy.org/job/OMERO-DEV-merge-push/579/). See the [console output](https://ci.openmicroscopy.org/job/OMERO-DEV-merge-push/579/consoleText) for more details.
Possible conflicts:
  - PR #5047 will-moore 'Api wells'
    - components/tools/OmeroWeb/test/integration/test_api_images.py
  - PR #5058 will-moore 'Api child count'
    - components/tools/OmeroWeb/test/integration/test_api_images.py

snoopycrimecop commented:
Conflicting PR. Removed from build [OMERO-DEV-merge-push#580](https://ci.openmicroscopy.org/job/OMERO-DEV-merge-push/580/). See the [console output](https://ci.openmicroscopy.org/job/OMERO-DEV-merge-push/580/consoleText) for more details.
Possible conflicts:
  - PR #5047 will-moore 'Api wells'
    - components/tools/OmeroWeb/test/integration/test_api_images.py
  - PR #5058 will-moore 'Api child count'
    - components/tools/OmeroWeb/test/integration/test_api_images.py

snoopycrimecop commented:
Conflicting PR. Removed from build [OMERO-DEV-merge-push#581](https://ci.openmicroscopy.org/job/OMERO-DEV-merge-push/581/). See the [console output](https://ci.openmicroscopy.org/job/OMERO-DEV-merge-push/581/consoleText) for more details.
Possible conflicts:
  - PR #5047 will-moore 'Api wells'
    - components/tools/OmeroWeb/test/integration/test_api_images.py
  - PR #5058 will-moore 'Api child count'
    - components/tools/OmeroWeb/test/integration/test_api_images.py

will-moore committed:
Tiny fix to test_api_images to avoid merge conflict

will-moore committed:
Merge remote-tracking branch 'origin/develop' into api_save_whitelist

will-moore committed:
Test handling of create, update & delete for unsupported types

will-moore committed:
Move test_crud_unsupported to test_api_containers.py

Since this test is really the counterpart to test_container_crud() above

will-moore committed:
Remove comment

will-moore committed:
flake8 fix

jburel commented on components/tools/OmeroWeb/omeroweb/api/decorators.py:
```diff
@@ -27,7 +27,8 @@
 from django.http import JsonResponse
 from functools import update_wrapper
-from api_exceptions import NotFoundError, BadRequestError, CreatedObject
+from api_exceptions import NotFoundError, BadRequestError, CreatedObject, \
```
one per line and alphabetical order

jburel commented on components/tools/OmeroWeb/omeroweb/api/decorators.py:
```diff
@@ -75,7 +76,7 @@ def handle_error(self, ex, trace):
         # But we try to handle all 'expected' errors appropriately
         # TODO: handle omero.ConcurrencyException
```
Will you do that in another PR? i.e. TODO

will-moore replied:
Yes. Hard to see that we need to support ConcurrencyException since we are doing very simple get/set queries now and no stateful services.

jburel commented on components/tools/OmeroWeb/omeroweb/api/views.py:
```diff
@@ -33,7 +33,7 @@
 from omeroweb.connector import Server
 from omeroweb.api.api_exceptions import BadRequestError, NotFoundError, \
-    CreatedObject
+    CreatedObject, MethodNotSupportedError
```
same

jburel commented on components/tools/OmeroWeb/test/integration/test_api_containers.py:
```diff
@@ -178,8 +179,36 @@ def user_screens(self, user1):
+    @pytest.mark.parametrize("dtype", ['Plate', 'Image', 'Well',
+                                       'Channel', 'foo'])
+    def test_crud_unsupported(self, user1, dtype):
+        """Test create, update & delete are rejected for unsupported types."""
```
I think you should split the test and not have ``post/put/delete`` tests in the same one

will-moore replied:
@jburel I split the test in 102c6c9

will-moore commented:
Need to fix failing ```test_api_errors.py```` tests in https://ci.openmicroscopy.org/view/Failing/job/OMERO-DEV-breaking-integration-Python27/164/testReport/
Haven't had time today, so will close this until I fix tests (tomorrow).

will-moore committed:
Fix failing test_api_errors.py tests

will-moore committed:
Tidy api_exceptions imports

will-moore committed:
Split test_crud_unsupported into separate tests

Single test for PUT, POST & DELETE is spilt into separate tests for each method.
To try and reduce code duplication, PUT and POST are covered by parametrizing method
in a single test.

will-moore committed:
flake8 fix

will-moore committed:
flake8 fix

will-moore commented:
@jburel After those last commits, this is going to conflict with #5058, so we'll need to get that merged ASAP. I still have 3 further api PRs waiting after this one.
@chris-allan You still have "requested changes" on this PR, which I think are addressed with the added tests?

snoopycrimecop commented:
Conflicting PR. Removed from build [OMERO-DEV-merge-push#592](https://ci.openmicroscopy.org/job/OMERO-DEV-merge-push/592/). See the [console output](https://ci.openmicroscopy.org/job/OMERO-DEV-merge-push/592/consoleText) for more details.
Possible conflicts:
  - PR #5058 will-moore 'Api child count'
    - components/tools/OmeroWeb/test/integration/test_api_errors.py

chris-allan commented on components/tools/OmeroWeb/omeroweb/api/api_exceptions.py:
```diff
@@ -50,6 +50,22 @@ def __init__(self, message, stacktrace=None):
+
+    status = 405
+
+    def __init__(self, message, stacktrace=None):
```
I'm not seeing any use of the `stacktrace` keyword argument in the implementation. Is it your intention to use it for something? If so, what? If you want to have a unified constructor prototype that is similar to the other exceptions you have above then I would suggest making a class hierarchy so you are not having to implement, and document, this every time.

chris-allan reviewed (changes requested):
Looks great aside from the exception feedback above.

will-moore committed:
ApiException base class to avoid code duplication

will-moore commented:
@chris-allan Thanks. Fixed in https://github.com/openmicroscopy/openmicroscopy/pull/5059/commits/eb18edcb38a5c89c68b7fa0d1f5221ed4af836d7

will-moore committed:
Merge remote-tracking branch 'origin/develop' into api_save_whitelist

jburel commented:
Exception code fixed.
Tests split and passing
I will merge on Monday am if everything is green
